File: vscodeExt/75.sort-colors.py
#
# @lc app=leetcode id=75 lang=python3
#
# [75] Sort Colors
#

# @lc code=start
class Solution:
    def sortColors(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        zero = -1 #[0,zero] are zeroes
        two = len(nums) # [two, len(nums)-1] are twos 
        i = 0 
        while i < two:
            if nums[i] == 0:
                zero += 1
                nums[zero], nums[i] = nums[i], nums[zero]
                i += 1
            elif nums[i] == 1:
                i += 1
            else: # num[i] == 2
                two -= 1
                nums[i], nums[two] = nums[two], nums[i]
                         

        
# @lc code=end

File: vscodeExt/75.sort-colors.java
/*
 * @lc app=leetcode id=75 lang=java
 *
 * [75] Sort Colors
 */

// @lc code=start
class Solution {
    public void sortColors(int[] nums) {
        // [0,index0) are 0s
        int index0 = 0;
        // (index2, nums.length-1] are 2s;
        int index2 = nums.length-1;
        for(int i =0;i<=index2;){
            if(nums[i]==0){
                int t = nums[index0];
                nums[index0] =0;
                nums[i] = t;
                index0++;
                i++;
            } else if (nums[i]==1){
                i++;
            } else {
                int t = nums[index2];
                nums[index2] =2;
                nums[i] = t;
                index2--;
            } 
        }
        
    }
}
// @lc code=end

